handler/MenuHandler.py

```diff
@@ -2,66 +2,157 @@
+	def __broadcast(cls, request: str):
+		""" Handle the peer request
+		Parameters:
+			request - the list containing the request parameters
+		"""
+		neighbours = AppData.get_neighbours()
+
+		for neighbour in neighbours:
+			sd = cls.__create_socket(neighbour[0], neighbour[1])
+			if sd is not None:
+				try:
+					sd.sendall(request.encode())
+					sd.close()
+				except socket.error:
+					print(f'Impossible to send data to {neighbour[0]} on port {neighbour[1]}\n')
+			else:
+				print(f'Cannot connect to {neighbour[0]} on port {neighbour[1]}\n')
+
+	@classmethod
+	def __unicast(cls, host: str, port: str, request: str):
+		sd = cls.__create_socket(host, port)
+		if sd is not None:
+			try:
+				sd.sendall(request.encode())
+				sd.close()
+			except socket.error:
+				print(f'Impossible to send data to {host} on port {port}\n')
+		else:
+			print(f'Cannot connect to {host} on port {port}\n')
 
-	def serve(self, choice: str) -> None:
+	@classmethod
+	def serve(cls, choice: str) -> None:
 		""" Handle the peer request
 		Parameters:
 			request - the list containing the request parameters
 		Returns:
 			str - the response
 		"""
+		print("Yeah! You chose: " + choice + "\n\n")
 
 		if choice == "QUER":
+
 			# codice che manda il pkt sulla socket
+			pktid = str(uuid.uuid4().hex[:16].upper())
+			ip = '172.016.001.001|FC00:2001:db8a:a0b2:12f0:a13w:0001:0001'
+			port = '4000'
+
+			search = input('\nEnter the file name: ')
+
+			# aggiungere verifica lunghezza file name
 
-			# avvio il server di ricezione delle response
-			p = Process(target=lambda: Server(4000, SelfHandler()).run())
-			p.daemon = True
-			p.start()
+			request = choice+pktid+ip+port+search
 
-			# avvio il timer
-			t = Timer(300, self.__kill_server, args=(p,))
+			# avvio il server di ricezione delle response, lo faccio prima del broadcast
+			# per evitare che i primi client che rispondono non riescano a connettersi
+			t = Thread(target=lambda: Server(4000, SelfHandler()).run(True))
+			t.daemon = True
 			t.start()
 
-			input('\nFile search in progress, press enter to continue...\n')
+			cls.__broadcast(request)
+			t.join()
 
-			if t.isAlive:
-				t.cancel()
-				self.__kill_server(p)
+			files = AppData.search_in_peer_files(search)
 
-			choice = input('Please select a peer:')
+			for count, file in enumerate(files, start=1):
+				print(f'{count}]', file)
 
-		elif choice == "NEAR":
-			pass
+			index = input('Please select a file to download:')
+			if files[int(index)][0]:
+				host = files[int(index)][0]
+			else:
+				host = files[int(index)][1]
+			port = files[int(index)][2]
+			file_md5 = files[int(index)][3]
+			file_name = files[int(index)][4]
 
-		elif choice == "RETR":
+			# preparo request per retr, faccio partire server in attesa download, invio request e attendo
+			request = 'RETR'+file_md5+file_name
 
-			if len(request) != 36:
-				return "Invalid request, usage is RETR<Filemd5>"
+			t = Thread(target=lambda: Server(4000, SelfHandler()).run(True))
```

Comment:
Questo fa sorgere una piccola problematica a cui non avevamo pensato. Se riapriamo il server sulla porta 4000 che abbiamo appena chiuso, c'è caso che possiamo ricevere altre AQUE relative alla ricerca che abbiamo appena deciso di interrompere. Mi sa che le ARET è meglio se le riceviamo su un' altra porta ancora.

Reply:
In realtà no, ripensandoci la ARET viene mandata in risposta sulla stessa socket che ha mandato la RETR. Noi stessi facciamo così. Questo ci evita la necessità di dover istanziare un altro Server.
Quindi basta fare una recv dopo la send di ```request = 'RETR'+file_md5+file_name```.

A questo punto il SelfHandler di @GiovanniFiorini non gestisce più le ARET, che quindi devono essere gestite dal MenuHandler come conseguenza di una RETR. Che dite?

Reply:
Secondo me è meglio tenere le due cose separate per mantenere incapsulato il server ed evitare di scrivere troppo codice, che per altro sarà per la maggior parte un doppione. A sto punto io voto di cambiare semplicemente porta, che è comodissimo così.

Reply:
Per lo stesso motivo ho deciso di cambiare la porta anche della NEAR-ANEA, perché esiste la possibilità che l'operazione sia talmente rapida da permettere a possibili pacchetti rimanenti in rete e diretti alla porta 4000 di raggiungere il server durante una ANEA. Avere una porta dedicata a ciascuna richiesta mi sembra una buona prassi, quindi ripeto, voto per quest'approccio.